### PR TITLE
[esp32_improv] Add triggers for various states

### PR DESCRIPTION
**Pull request by kbx81**
# What does this implement/fix?

SSIA

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#4256

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/esp32_improv/esp32_improv_component.cpp [hunk 1]
```diff
@@ -68,7 +68,12 @@ void ESP32ImprovComponent::setup_characteristics() {
 
 void ESP32ImprovComponent::loop() {
   if (!global_ble_server->is_running()) {
-    this->state_ = improv::STATE_STOPPED;
+    if (this->state_ != improv::STATE_STOPPED) {
+      this->state_ = improv::STATE_STOPPED;
+#ifdef USE_ESP32_IMPROV_STATE_CALLBACK
+      this->state_callback_.call(this->state_);
+#endif
```
Comment (nielsnl68) [inline esphome/components/esp32_improv/esp32_improv_component.cpp]:
```suggestion
      this->state_callback_.call(this->state_);
```

File: esphome/components/esp32_improv/esp32_improv_component.cpp [hunk 2]
```diff
@@ -217,6 +222,9 @@ void ESP32ImprovComponent::set_state_(improv::State state) {
   service_data[7] = 0x00;  // Reserved
 
   esp32_ble::global_ble->advertising_set_service_data(service_data);
+#ifdef USE_ESP32_IMPROV_STATE_CALLBACK
+  this->state_callback_.call(state);
+#endif
```
Comment (nielsnl68) [inline esphome/components/esp32_improv/esp32_improv_component.cpp]:
```suggestion
  this->state_callback_.call(state);
```

File: esphome/components/esp32_improv/esp32_improv_component.h [hunk 3]
```diff
@@ -42,6 +46,12 @@ class ESP32ImprovComponent : public Component, public BLEServiceComponent {
   void stop() override;
   bool is_active() const { return this->state_ != improv::STATE_STOPPED; }
 
+#ifdef USE_ESP32_IMPROV_STATE_CALLBACK
+  void add_on_state_callback(std::function<void(improv::State)> &&callback) {
+    this->state_callback_.add(std::move(callback));
+  }
+#endif
```
Comment (nielsnl68) [inline esphome/components/esp32_improv/esp32_improv_component.h]:
```suggestion
  void add_on_state_callback(std::function<void(improv::State, improv::Error)> &&callback) {
    this->state_callback_.add(std::move(callback));
  }
```
Optional Allow please sending the error state to the events as well.

File: esphome/components/esp32_improv/esp32_improv_component.h [hunk 4]
```diff
@@ -84,6 +96,9 @@ class ESP32ImprovComponent : public Component, public BLEServiceComponent {
 
   improv::State state_{improv::STATE_STOPPED};
   improv::Error error_state_{improv::ERROR_NONE};
+#ifdef USE_ESP32_IMPROV_STATE_CALLBACK
+  CallbackManager<void(improv::State)> state_callback_{};
+#endif
```
Comment (nielsnl68) [inline esphome/components/esp32_improv/esp32_improv_component.h]:
```suggestion
  CallbackManager<void(improv::State)> state_callback_{};
```

File: esphome/components/esp32_improv/esp32_improv_component.h [hunk 5]
```diff
@@ -54,6 +64,8 @@ class ESP32ImprovComponent : public Component, public BLEServiceComponent {
   void set_wifi_timeout(uint32_t wifi_timeout) { this->wifi_timeout_ = wifi_timeout; }
   uint32_t get_wifi_timeout() const { return this->wifi_timeout_; }
 
+  improv::State get_improv_state() const { return this->state_; }
+
```
Comment (nielsnl68) [inline esphome/components/esp32_improv/esp32_improv_component.h:68]:
```suggestion
   improv::State get_improv_error_state() const { return this->error_; }
   
```
Allow external access to the latest error state.

File: esphome/components/esp32_improv/__init__.py [hunk 6]
```diff
@@ -63,3 +130,31 @@ async def to_code(config):
...
+    for conf in config.get(CONF_ON_STATE, []):
+        trigger = cg.new_Pvariable(conf[CONF_TRIGGER_ID], var)
+        await automation.build_automation(trigger, [(State, "state")], conf)
+        use_state_callback = True
+    for conf in config.get(CONF_ON_STOPPED, []):
+        trigger = cg.new_Pvariable(conf[CONF_TRIGGER_ID], var)
+        await automation.build_automation(trigger, [], conf)
+        use_state_callback = True
+    if use_state_callback:
+        cg.add_define("USE_ESP32_IMPROV_STATE_CALLBACK")
```
Comment (nielsnl68) [inline esphome/components/esp32_improv/__init__.py:146]:
```suggestion
      cg.add_define("USE_ESP32_IMPROV")
```

File: esphome/components/esp32_improv/esp32_improv_component.h [hunk 7]
```diff
@@ -9,6 +9,10 @@
 #include "esphome/components/esp32_ble_server/ble_server.h"
 #include "esphome/components/wifi/wifi_component.h"
 
+#ifdef USE_ESP32_IMPROV_STATE_CALLBACK
+#include "esphome/core/automation.h"
+#endif
```
Comment (nielsnl68) [inline esphome/components/esp32_improv/esp32_improv_component.h:14]:
```suggestion
#include "esphome/core/automation.h"
```

**Review by nielsnl68 — CHANGES_REQUESTED** [review]
I nice addition to the improv compronent, just on request to drop the **USE_ESP32_IMPROV_STATE_CALLBACK** define in favor of a global **USE_ESP32_IMPROV**

**Comment by kbx81** [comment]
> I nice addition to the improv compronent, just on request to drop the **USE_ESP32_IMPROV_STATE_CALLBACK** define in favor of a global **USE_ESP32_IMPROV**

@nielsnl68 I talked with Jesse about this and we will keep the `#defines`. The BT stack itself already significantly increases the size of the binary and we want to exclude code where it's not in use (which, for the new code in this PR, will be everywhere, at least initially).

If a specific component requires the callbacks, it can simply `cg.add_define()` [as the `esp32_ble_tracker` does for OTA](https://github.com/esphome/esphome/blob/5a3e1d57921a839a5f383350e45cdd3304efa311/esphome/components/esp32_ble_tracker/__init__.py#L275).

**Review by jesserockz — APPROVED** [review]

**Pull request merged**